ra/ra.go

```diff
@@ -1376,24 +1376,37 @@ func (ra *RegistrationAuthorityImpl) checkCertificatesPerFQDNSetLimit(ctx contex
 		return nil
 	}
 
-	issuanceTimestampsInWindow, err := ra.SA.FQDNSetTimestampsForWindow(ctx, &sapb.CountFQDNSetsRequest{
+	prevIssuances, err := ra.SA.FQDNSetTimestampsForWindow(ctx, &sapb.CountFQDNSetsRequest{
 		Domains: names,
 		Window:  limit.Window.Duration.Nanoseconds(),
 	})
 	if err != nil {
 		return fmt.Errorf("checking duplicate certificate limit for %q: %s", names, err)
 	}
 
-	if int64(len(issuanceTimestampsInWindow.Timestamps)) >= threshold {
-		// FQDNSetTimestampsForWindow returns the issuance timestamps in
-		// ascending order so we can assume that the first one is the oldest.
-		retryAfter := time.Unix(0, issuanceTimestampsInWindow.Timestamps[0]).Add(limit.Window.Duration)
+	if int64(len(prevIssuances.Timestamps)) < threshold {
+		// Issuance in window is below the threshold, no need to limit.
+		return nil
+	} else {
+		// Evaluate the rate limit using a leaky bucket algorithm. The bucket
+		// has a capacity of threshold and is refilled at a rate of 1 token per
+		// limit.Window/threshold from the time of each issuance timestamp.
+		now := ra.clk.Now()
+		nsPerToken := limit.Window.Nanoseconds() / threshold
+		for i, timestamp := range prevIssuances.Timestamps {
+			tokenGeneratedAt := now.Add(-time.Duration(int64(1*i+1) * nsPerToken))
```

Comment:
```suggestion
			tokenGeneratedAt := now.Add(-time.Duration(int64(i+1) * nsPerToken))
```

`1*i` is odd. :-)

Also I think the current logic is correct, but the naming is a bit off. A token wasn't notionally generated "at" the moment we're calculating here - that would indicate tokens were generated at query time, which is not right.

Instead, we know a token was generated _since_ the moment we're calculating. We don't know the complete status of the bucket at this point, we just know that within the last `nsPerToken` nanoseconds, a token was generated; if there was no issuance during that same interval, we have at least one token that can be spent.

I think it would be better as `tokensGeneratedSince`, and the logic below can be flipped:

```
issuanceTimestamp := time.Unix(0, timestamp)
if issuanceTimestamp.Before(tokensGeneratedSince) {
  // We know `i+1` tokens were generated since `tokenGeneratedSince`, and only `i` certificates were issued, so there's room in the rate limit.
  return nil
}
```

Reply:
> 1*i is odd. :-)
🤦🏻‍♀️